File: backend/app/db/models/study_session.py
import datetime
import uuid

from sqlalchemy import UUID, DateTime, ForeignKey, Index, Integer, text
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base, TimestampMixin


class StudySession(Base, TimestampMixin):
    __tablename__ = "study_sessions"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, server_default=text("uuidv7()")
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False
    )
    topic_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("topics.id", ondelete="CASCADE"), nullable=False
    )
    start_time: Mapped[datetime.datetime] = mapped_column(DateTime)
    end_time: Mapped[datetime.datetime] = mapped_column(DateTime)
    duration_minutes: Mapped[int] = mapped_column(Integer)

    __table_args__ = (Index("idx_study_sessions_user", "user_id"),)

File: personal_study_tracker/lib/features/sessions/domain/entities/study_session.dart
class StudySession {
  final String id;
  final String topicId;
  final DateTime startTime;
  final DateTime? endTime;
  final int durationMinutes;

  StudySession({
    required this.id,
    required this.topicId,
    required this.startTime,
    this.endTime,
    this.durationMinutes = 0,
  });

  bool get isActive => endTime == null;
}
